### PR TITLE
fix: can't backup database when EAR is enabled

### PR DESCRIPTION
**Pull request by johnxnguyen**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues
If EAR is enabled, then trying to create a database backup will fail.

### Causes
Since EAR is enabled, the database must first be migrated away from EAR before it can be backed up. This flow was still using the old `EncryptionKeys`, so the database key was not available.

### Solutions
- Delete the now retired `EncryptionKeys` and it's usage.
- Pass in the database key when creating a backup.

#### Test Coverage 

- Updated `CoreDataStackTests_Backup`

### Notes

Since we deleted the `EncryptionKeys`, we also needed update several tests. This also includes porting over security tests.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift [hunk 1]
```diff
@@ -126,12 +96,12 @@ extension NSManagedObjectContext {
...
             saveOrRollback()
         } catch {
-            WireLogger.ear.error("failed to migrate existing data toward EAR: \(error)")
-            reset()
             throw error
         }
     }
 
+            WireLogger.ear.error("failed to migrate existing data toward EAR: \(error)")
+            reset()
```
Comment (David-Henner) [inline wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift]: this probably shouldn't be here, I guess it should be in the `catch`

File: wire-ios-data-model/Source/Utilis/EAR/EARService.swift [hunk 2]
```diff
@@ -168,6 +168,16 @@ public class EARService: EARServiceInterface {
         }
     }
 
+    private var isEAREnabled: Bool {
```
Comment (netbe) [inline wire-ios-data-model/Source/Utilis/EAR/EARService.swift]: I would see it more as a function as it is not cheap to do (perform on all contexts)

File: wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift [hunk 3]
```diff
@@ -150,42 +120,12 @@ extension NSManagedObjectContext {
...
-            try migrateInstancesAwayFromEncryptionAtRest(type: ZMClientMessage.self, key: encryptionKeys.databaseKey)
-            try migrateInstancesAwayFromEncryptionAtRest(type: ZMConversation.self, key: encryptionKeys.databaseKey)
-        } catch {
-            encryptMessagesAtRest = true
             throw error
         }
     }
 
+            WireLogger.ear.error("failed to migrate existing data away from EAR: \(error)")
+            reset()
```
Comment (David-Henner) [inline wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift]: same here

File: wire-ios-data-model/Source/Utilis/EAR/EARService.swift (hunk 2 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/Utilis/EAR/EARService.swift]: done

File: wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift [hunk 4]
```diff
@@ -126,12 +96,10 @@ extension NSManagedObjectContext {
             )
             saveOrRollback()
         } catch {
-            WireLogger.ear.error("failed to migrate existing data toward EAR: \(error)")
-            reset()
```
Comment (David-Henner) [inline wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift]: should this be deleted?

File: wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift (hunk 1 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift]: Yeah somehow the merge got a bit messed up, already fixed

File: wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift [hunk 5]
```diff
@@ -150,42 +118,10 @@ extension NSManagedObjectContext {
             )
             saveOrRollback()
         } catch {
-            WireLogger.ear.error("failed to migrate existing data away from EAR: \(error)")
-            reset()
```
Comment (David-Henner) [inline wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift]: same here

Comment (johnxnguyen) [reply wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift]: Good catch, fixed

File: wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift (hunk 4 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift]: good catch, fixed

**Review by David-Henner — APPROVED** [review]

**Review by netbe — APPROVED** [review]

**Pull request merged**